File: 104. Maximum Depth of Binary Tree/solution_dfs_iterative.py
from typing import *

class TreeNode:
    """Definition for a binary tree node."""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        
        stack = [(root, 1)] # (node, depth)
        maxDepth = 0

        while stack:
            node, depth = stack.pop()
        
            if node:
                maxDepth = max(maxDepth, depth)
        
                stack.append((node.left, depth + 1))
                stack.append((node.right, depth + 1))
        
        return maxDepth

File: 104. Maximum Depth of Binary Tree/solution_dfs_iterative.js
/** Definition for a binary tree node. */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function (root) {
    
    const stack = [[root, 1]]; // [node, depth]
    let maximumDepth = 0;

    while (stack.length > 0) {
        const [node, depth] = stack.pop();

        if (node) {
            maximumDepth = Math.max(maximumDepth, depth);

            stack.push([node.left, depth + 1]);
            stack.push([node.right, depth + 1]);
        }
    }

    return maximumDepth;
};